Security_Engine/Compute/CameraFieldOfView.cs

```diff
@@ -0,0 +1,351 @@
+                {
+                    Base.Compute.RecordWarning("Camera Device is inside obstacle. Null value will be returned.");
+                    return null;
+                }
+                List<Polyline> intersection = obstacle.BooleanIntersection(cameraConePolyline);
+                if (intersection.Count != 0)
+                {
+                    intersectObstacles.Add(obstacle);
+                    foreach (Polyline intersectPolyline in intersection)
+                    {
+                        List<Point> points = intersectPolyline.ControlPoints;
+                        intersectPoints.AddRange(points);
+                    }
+                }
+            }
+            intersectPoints = intersectPoints.CullDuplicates();
+
+            //create and sort extended ray lines passing through intersected points
+            List<Line> rayLines = new List<Line>();
+            foreach (Point point in intersectPoints)
+            {
+                Line line = BH.Engine.Geometry.Create.Line(cameraLocation, point);
+                Line rayLine = line.Extend(0, radius - line.Length());
+                rayLines.Add(rayLine);
+            }
+            rayLines = rayLines.CullDuplicateLines();
+            rayLines = rayLines.OrderBy(x => x.SingedAngle(rayLines[0], Vector.ZAxis)).ToList();
+
+            //split ray lines and find visible line
+            List<Dictionary<Line, Polyline>> linesDict = new List<Dictionary<Line, Polyline>>();
+            foreach (Line rayLine in rayLines)
+            {
+                Line visibleLine = rayLine.VisibleLine(intersectObstacles);
+                linesDict.Add(visibleLine.LineObstacleDictionary(intersectObstacles));
+            }
+
+            //create points chain
+            List<Point> pointsChain = linesDict.PointsChain(cameraLocation, radius, distanceTolerance);
+
+            //create cone
+            Arc coneArc = cameraCone.Curves[1] as Arc;
+            PolyCurve cameraViewPolyCurve = pointsChain.CameraViewPolyCurve(cameraLocation, radius, coneArc, distanceTolerance);
+
+            return cameraViewPolyCurve;
+        }
+
+        /***************************************************/
+
+        [Description("Project obstacle polyline on camera plane.")]
+        [Input("obstacle", "Obstacle to project.")]
+        [Input("cameraPlane", "Plane to project the obstacle on.")]
+        [Output("newObstacle", "Obstacle polyline projected on camera plane.")]
+
+        private static Polyline FixAndProject(this Polyline obstacle, Plane cameraPlane, double tolerance)
+        {
+            if (obstacle.IsInPlane(cameraPlane))
+                return obstacle;
+
+            if (!obstacle.IsClosed(tolerance) || !obstacle.IsPlanar(tolerance))
+            {
+                Base.Compute.RecordWarning("Obstacle polyline is open and not planar. It will be closed and projected on camera plane.");
```

Comment:
```suggestion
                Base.Compute.RecordWarning("Obstacle polyline is open or not planar. It will be closed and projected on camera plane.");
```